core/kernel/transfer_list.c

```diff
@@ -0,0 +1,524 @@
+	transfer_list_unmap_sync(tl);
+
+	return new_tl;
+}
+
+/*******************************************************************************
+ * Verifying the header of a transfer list
+ * Compliant to 2.4.1 of Firmware handoff specification (v0.9)
+ * Return transfer list operation status code
+ ******************************************************************************/
+int transfer_list_check_header(const struct transfer_list_header *tl)
+{
+	if (!tl)
+		return TL_OPS_NON;
+
+	if (tl->signature != TRANSFER_LIST_SIGNATURE) {
+		EMSG("Bad transfer list signature %#"PRIx32"",
+		     tl->signature);
+		return TL_OPS_NON;
+	}
+
+	if (!tl->max_size) {
+		EMSG("Bad transfer list max size %#"PRIx32"",
+		     tl->max_size);
+		return TL_OPS_NON;
+	}
+
+	if (tl->size > tl->max_size) {
+		EMSG("Bad transfer list size %#"PRIx32"", tl->size);
+		return TL_OPS_NON;
+	}
+
+	if (tl->hdr_size != sizeof(struct transfer_list_header)) {
+		EMSG("Bad transfer list header size %#"PRIx8"",
+		     tl->hdr_size);
+		return TL_OPS_NON;
+	}
+
+	if (!transfer_list_verify_checksum(tl)) {
+		EMSG("Bad transfer list checksum %#"PRIx8"", tl->checksum);
+		return TL_OPS_NON;
+	}
+
+	if (tl->version == 0) {
+		EMSG("Transfer list version is invalid");
+		return TL_OPS_NON;
+	} else if (tl->version == TRANSFER_LIST_VERSION) {
+		IMSG("Transfer list version is valid for all operations");
+		return TL_OPS_ALL;
+	} else if (tl->version > TRANSFER_LIST_VERSION) {
+		IMSG("Transfer list version is valid for read-only");
+		return TL_OPS_RO;
+	}
+
+	IMSG("Old transfer list version is detected");
+	return TL_OPS_CUS;
+}
+
+/*******************************************************************************
+ * Enumerate the next transfer entry
+ * Return pointer to the next transfer entry or NULL on error
+ ******************************************************************************/
+struct transfer_list_entry *transfer_list_next(struct transfer_list_header *tl,
+					       struct transfer_list_entry *last)
+{
+	struct transfer_list_entry *te = NULL;
+	vaddr_t tl_ev = 0;
+	vaddr_t va = 0;
+	vaddr_t ev = 0;
+	size_t sz = 0;
+
+	if (!tl)
+		return NULL;
+
+	tl_ev = (vaddr_t)tl + tl->size;
+
+	if (last) {
+		va = (vaddr_t)last;
+		/* check if the total size overflow */
+		if (ADD_OVERFLOW(last->hdr_size, last->data_size, &sz))
+			return NULL;
+		/* roundup to the next entry */
+		if (ADD_OVERFLOW(va, sz, &va) ||
+		    ROUNDUP_OVERFLOW(va, TRANSFER_LIST_GRANULE, &va))
+			return NULL;
+	} else {
+		va = (vaddr_t)tl + tl->hdr_size;
+	}
+
+	te = (struct transfer_list_entry *)va;
+
+	if (va + sizeof(*te) > tl_ev || te->hdr_size < sizeof(*te) ||
+	    ADD_OVERFLOW(te->hdr_size, te->data_size, &sz) ||
+	    ADD_OVERFLOW(va, sz, &ev) ||
+	    ev > tl_ev)
+		return NULL;
+
+	return te;
+}
+
+/*******************************************************************************
+ * Calculate the byte sum of a transfer list
+ * Return byte sum of the transfer list
```

Comment:
Might be worth mentioning that this is modulo 256.

Reply:
Fixed in the fixup patches.